python/spear/sp_engine.py

```diff
@@ -0,0 +1,271 @@
+        elif sys.platform == "linux":
+            assert launch_executable_ext == ".sh"
+            launch_executable_internal = launch_executable
+        else:
+            assert False
+
+        assert os.path.exists(launch_executable_internal)
+
+        launch_args.append("-windowed")
+        launch_args.append("-resx={}".format(self._config.SPEAR.WINDOW_RESOLUTION_X))
+        launch_args.append("-resy={}".format(self._config.SPEAR.WINDOW_RESOLUTION_Y))
+        launch_args.append("-graphicsadapter={}".format(self._config.SPEAR.GPU_ID))
+        launch_args.append("-nosound")
+        # launch_args.append("-fileopenlog")         # generate a log of which files are opened in which order
+        launch_args.append("-stdout")              # ensure log output is written to the terminal 
+        launch_args.append("-fullstdoutlogoutput") # ensure log output is written to the terminal
+        launch_args.append("-nologtimes")          # don't print timestamps next to log messages twice
+
+        if self._config.SPEAR.RENDER_OFFSCREEN:
+            launch_args.append("-renderoffscreen")
+
+        if self._config.SPEAR.UNREAL_INTERNAL_LOG_FILE != "":
+            launch_args.append("-log={}".format(self._config.SPEAR.UNREAL_INTERNAL_LOG_FILE))
+       
+        launch_args.append("-config_file={}".format(temp_config_file))
+
+        for a in self._config.SPEAR.CUSTOM_COMMAND_LINE_ARGUMENTS:
+            launch_args.append("{}".format(a))
+
+        cmd = [launch_executable_internal] + launch_args
+
+        spear.log("Launching executable with the following command-line arguments:")
+        spear.log_no_prefix(" ".join(cmd))
+
+        spear.log("Launching executable with the following config values:")
+        spear.log_no_prefix(self._config)
+
+        popen = Popen(cmd)
+        self._process = psutil.Process(popen.pid)
+
+        # see https://github.com/giampaolo/psutil/blob/master/psutil/_common.py for possible status values
+        status = self._process.status()
+        if status not in ["running", "sleeping", "disk-sleep"]:
+            spear.log("ERROR: Unrecognized process status: " + status)
+            spear.log("ERROR: Killing process " + str(self._process.pid) + "...")
+            self._force_kill_unreal_instance()
+            self._close_rpc_client()
+            assert False
+
+    def _request_close_unreal_instance(self):
+
+        if self._config.SPEAR.LAUNCH_MODE == "running_instance":
+            spear.log('SPEAR.LAUNCH_MODE == "running_instance" so we assume that the Unreal instance should remain open...')
+            return
+
+        spear.log("Closing Unreal instance...")
+
+        self._rpc_client.call("sp_engine.request_close")
+        self._wait_until_unreal_instance_is_closed()
+
+        spear.log("Finished closing Unreal instance.")
+
+    def _initialize_rpc_client(self):
```

Comment:
```
    def _initialize_rpc_client(self):

        spear.log("Initializing RPC client...")

        connected = False
        
        # if we're connecting to a running instance, then we assume that the RPC server is already running and only try to connect once
        if self._config.SPEAR.INSTANCE.LAUNCH_MODE == "none":

            try:
                self._rpc_client = msgpackrpc.Client(
                    msgpackrpc.Address("127.0.0.1", self._config.SP_ENGINE.PORT),
                    timeout=self._config.SPEAR.INSTANCE.RPC_CLIENT_INTERNAL_TIMEOUT_SECONDS,
                    reconnect_limit=self._config.SPEAR.INSTANCE.RPC_CLIENT_INTERNAL_RECONNECT_LIMIT)
                self._rpc_client.call("sp_engine.ping")
                connected = True

            except Exception as e:
                # Client may not clean up resources correctly in this case, so we clean things up explicitly.
                # See https://github.com/msgpack-rpc/msgpack-rpc-python/issues/14 for more details.
                spear.log("Exception: ", e)
                self._close_rpc_client()

        # otherwise try to connect repeatedly, since the RPC server might not have started yet
        elif self._config.SPEAR.INSTANCE.LAUNCH_MODE in ["editor", "standalone"]:

            start_time_seconds = time.time()
            elapsed_time_seconds = time.time() - start_time_seconds
            while elapsed_time_seconds < self._config.SPEAR.INSTANCE.INITIALIZE_RPC_CLIENT_MAX_TIME_SECONDS:

                # see https://github.com/giampaolo/psutil/blob/master/psutil/_common.py for possible status values
                status = self._process.status()
                if status not in ["disk-sleep", "running", "sleeping"]:
                    spear.log("ERROR: Unexpected process status: " + status)
                    break

                try:
                    self._rpc_client = msgpackrpc.Client(
                        msgpackrpc.Address("127.0.0.1", self._config.SP_ENGINE.PORT), 
                        timeout=self._config.SPEAR.INSTANCE.RPC_CLIENT_INTERNAL_TIMEOUT_SECONDS, 
                        reconnect_limit=self._config.SPEAR.INSTANCE.RPC_CLIENT_INTERNAL_RECONNECT_LIMIT)
                    self._rpc_client.call("sp_engine.ping")
                    connected = True
                    break

                except Exception as e:
                    spear.log("Exception:", e)
                    self._close_rpc_client()

                time.sleep(self._config.SPEAR.INSTANCE.INITIALIZE_RPC_CLIENT_SLEEP_TIME_SECONDS)
                elapsed_time_seconds = time.time() - start_time_seconds

        else:
            assert false

        if not connected:
            spear.log("ERROR: Couldn't connect to RPC client, giving up...")
            if self._config.SPEAR.LAUNCH_MODE in ["editor", "standalone"]:
                self._force_kill_unreal_instance()
            assert False

        spear.log("Finished initializing RPC client.")
```